design/apidsl/security.go

```diff
@@ -0,0 +1,308 @@
+package apidsl
+
+import (
+	"github.com/goadesign/goa/design"
+	"github.com/goadesign/goa/dslengine"
+)
+
+// Security defines an authentication method required to access a goa
+// Action.  When defined on a Resource, it applies to all Actions,
+// unless overriden by individual actions.  When defined at the API
+// level, it will apply to all resources by default, following the
+// same logic.
+//
+// The method name refers to what you have passed to the different
+// OAuth2Security(), BasicAuthSecurity() and APIKeySecurity()
+// declarations.
+func Security(method string, dsl ...func()) {
+	def := &design.SecurityDefinition{Method: method}
+
+	if len(dsl) != 0 {
+		if !dslengine.Execute(dsl[0], def) {
+			return
+		}
+	}
+
+	parentDef := dslengine.CurrentDefinition()
+	switch parent := parentDef.(type) {
+	case *design.ActionDefinition:
+		parent.Security = def
+	case *design.ResourceDefinition:
+		parent.Security = def
+	case *design.APIDefinition:
+		parent.Security = def
+	default:
+		dslengine.IncompatibleDSL()
+		return
+	}
+}
+
+// NoSecurity resets the authentication methods for an Action or a
+// Resource. It also prevents fallback to Resource or API-defined
+// Security().
+func NoSecurity() {
+	def := &design.SecurityDefinition{NoSecurity: true}
+
+	parentDef := dslengine.CurrentDefinition()
+	switch parent := parentDef.(type) {
+	case *design.ActionDefinition:
+		parent.Security = def
+	case *design.ResourceDefinition:
+		parent.Security = def
+	default:
+		dslengine.IncompatibleDSL()
+		return
+	}
+}
+
+// BasicAuthSecurity defines a "basic" security method for the API.
+//
+// Example:
+//
+//     BasicAuthSecurity("password", func() {
+//         Description("Use your own password!")
+//     })
+//
+func BasicAuthSecurity(name string, dsl ...func()) {
+	if _, ok := dslengine.CurrentDefinition().(*design.APIDefinition); !ok {
```

Comment:
Again, this would be in order to reuse the thing as a var, returned by the `*Security()`, right ?

I can make sure it's in before merge if you like. Would we keep the option of having it inside API too ?